core/commands/ls.go

```diff
@@ -71,14 +81,20 @@ The JSON output contains type information.
 			return
 		}
 
+		rtype, rf, _ := req.Option("type").String()
+		if rf && rtype != resolveAlways && rtype != resolveLocal && rtype != resolveNever {
+			res.SetError(errInvalidResolveType, cmdkit.ErrClient)
+			return
+		}
+
 		resolve, _, err := req.Option("resolve-type").Bool()
 		if err != nil {
 			res.SetError(err, cmdkit.ErrNormal)
 			return
 		}
 
 		dserv := nd.DAG
-		if !resolve {
+		if rtype == resolveLocal || !resolve {
```

Comment:
This is confusing.  I would not use resolve and just test on `rtype`.  See above comment.